types/src/state_store/state_value.rs

```diff
@@ -15,48 +15,150 @@ use once_cell::sync::OnceCell;
 #[cfg(any(test, feature = "fuzzing"))]
 use proptest::{arbitrary::Arbitrary, prelude::*};
 use serde::{Deserialize, Deserializer, Serialize, Serializer};
-#[derive(
-    BCSCryptoHash,
-    Clone,
-    CryptoHasher,
-    Debug,
-    Deserialize,
-    Eq,
-    PartialEq,
-    Serialize,
-    Ord,
-    PartialOrd,
-    Hash,
-)]
-pub enum StateValueMetadata {
+
+#[derive(Deserialize, Serialize)]
+#[serde(rename = "StateValueMetadata")]
+pub enum PersistedStateValueMetadata {
     V0 {
         deposit: u64,
         creation_time_usecs: u64,
     },
+    V1 {
+        slot_deposit: u64,
+        bytes_deposit: u64,
+        creation_time_usecs: u64,
+    },
 }
 
-// To avoid nested options when fetching a resource and its metadata.
-pub type StateValueMetadataKind = Option<StateValueMetadata>;
+impl PersistedStateValueMetadata {
+    pub fn into_in_mem_form(self) -> StateValueMetadata {
+        match self {
+            PersistedStateValueMetadata::V0 {
+                deposit,
+                creation_time_usecs,
+            } => StateValueMetadata::new_impl(deposit, 0, creation_time_usecs),
+            PersistedStateValueMetadata::V1 {
+                slot_deposit,
+                bytes_deposit,
+                creation_time_usecs,
+            } => StateValueMetadata::new_impl(slot_deposit, bytes_deposit, creation_time_usecs),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+struct StateValueMetadataInner {
+    slot_deposit: u64,
+    bytes_deposit: u64,
+    creation_time_usecs: u64,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct StateValueMetadata {
+    inner: Option<StateValueMetadataInner>,
+}
 
 impl StateValueMetadata {
-    pub fn new(deposit: u64, creation_time_usecs: &CurrentTimeMicroseconds) -> Self {
-        Self::V0 {
-            deposit,
-            creation_time_usecs: creation_time_usecs.microseconds,
-        }
+    pub fn into_persistable(self) -> Option<PersistedStateValueMetadata> {
+        self.inner.map(|inner| {
+            let StateValueMetadataInner {
+                slot_deposit,
+                bytes_deposit,
+                creation_time_usecs,
+            } = inner;
+            if bytes_deposit == 0 {
```

Comment:
it's fine here, but for the future, if you add V2, you might want to actually have an enum in memory?

or you intentionally want to convert into newest possible?

Reply:
intentionally, because I feel better doing `Eq` when there's not two vairants representing the same thing.